SETUP/upgrade/21/20240414_update_current_tallies.php

```diff
@@ -0,0 +1,148 @@
+
+// iterate through all tallyboards
+foreach (get_all_current_tallyboards() as $tallyboard) {
+    $tally_name = $tallyboard->tally_name;
+    $holder_type = $tallyboard->holder_type;
+    echo "Populating for $holder_type, $tally_name\n";
+
+    // iterate over all of our holders, this should be fast because of the index
+    $sql = sprintf(
+        "
+        SELECT holder_id, tally_value, tally_delta, timestamp
+        FROM past_tallies
+        WHERE tally_name = '%s' AND holder_type = '%s' AND timestamp = %d
+        ORDER BY holder_id
+        ",
+        DPDatabase::escape($tally_name),
+        DPDatabase::escape($holder_type),
+        get_time_of_latest_snapshot($tallyboard)
+    );
+    $res = DPDatabase::query($sql);
+    $progress_index = 0;
+    while ($row = mysqli_fetch_assoc($res)) {
+        $holder_id = intval($row['holder_id']);
+
+        if ($progress_index % 100 == 0) {
+            echo sprintf("    %s: holder_id %d\n", $progress_index, $holder_id);
+        }
+
+        $sql = sprintf(
+            "
+            UPDATE current_tallies
+            SET
+                last_snap_tally_value = %d,
+                last_snap_tally_delta = %d
+            WHERE
+                tally_name = '%s'
+                AND holder_type = '%s'
+                AND holder_id = %d
+            ",
+            $row['tally_value'],
+            $row['tally_delta'],
+            DPDatabase::escape($tally_name),
+            DPDatabase::escape($holder_type),
+            $row['holder_id']
+        );
+        DPDatabase::query($sql);
+        $progress_index += 1;
+    }
+
+    echo "    Finding distinct timestamps...\n";
+    $sql = sprintf(
+        "
+        SELECT DISTINCT timestamp
+        FROM past_tallies
+        WHERE tally_name = '%s' AND holder_type = '%s'
+        ",
+        DPDatabase::escape($tally_name),
+        DPDatabase::escape($holder_type),
+    );
+    $res = DPDatabase::query($sql);
+
+    echo "    Inserting them into tally_snapshot_times...\n";
+    while ($row = mysqli_fetch_assoc($res)) {
+        $timestamp = $row['timestamp'];
+
+        // we handle duplicates in case this script was restarted
+        $sql = sprintf(
+            "
+            INSERT INTO tally_snapshot_times
+            SET
+                tally_name = '%s',
+                holder_type = '%s',
+                timestamp = %d
+            ON DUPLICATE KEY UPDATE
+                tally_name = tally_name,
+                holder_type = holder_type,
+                timestamp = timestamp
+            ",
+            DPDatabase::escape($tally_name),
+            DPDatabase::escape($holder_type),
+            $timestamp
+        );
+        DPDatabase::query($sql);
+    }
+}
+
+// ------------------------------------------------------------
+
+echo "\nDone!\n";
+
+// We have to use our own function in the upgrade script rather than the one
+// in Tallyboard because it will already be trying to use the table we're
+// trying to populate!
```

Comment:
> But it doesn't apply to the pgdp deployment plan you outline https://github.com/DistributedProofreaders/dproofreaders/pull/1199#discussion_r1600383854, because we'll be running the upgrade scripts before the new `TallyBoard.inc` is installed, right?

Correct. But I didn't want to assume that's how other sites might be doing it.

> On the other hand, what if (for each tallyboard) we populate tally_snapshot_times before the new columns in current_tallies? Then we could use Tallyboard::get_time_of_latest_snapshot, and drop this extra function. I think I like that better.

I see what you mean, but I like being explicit about what's going on in the upgrade script. Its possible that we, or others, might do something out of order / run something manually / need to rerun pieces out of order / etc and having a function that always works vs works based on half of the upgrade script seems safer overall.

I'll update the comment above it though, as your version is clearer.

Reply:
> Its possible that we, or others, might do something out of order / run something manually / need to rerun pieces out of order / etc and having a function that always works vs works based on half of the upgrade script seems safer overall.

Okay in general, but could you give a specific example of something that we or others might do that would cause my suggested change to malfunction (that wouldn't also cause the existing PR code to malfunction)?

---

Also, if we're recognizing the possibility that someone might need to rerun this script, then it seems like the `ALTER TABLE` and `CREATE TABLE` code at the start should allow for that. Currently, the script will die with an error if you re-execute either of those, won't it?

---

Are we allowing for the possibility that a snapshot might be running at the same time as this script? I'm pretty sure we aren't. We should maybe have a comment at the top saying something like:
`past_tallies` must remain unchanged over the run (+ possible reruns) of this script, so choose a time well away from any snapshot (or any ad hoc query that could change `past_tallies`).

On second thought, we should add more stuff to `SETUP/UPGRADE.md` (not necessarily this PR):
- Disable general access to your site. (Maybe put it in maintenance mode.)
- Disable cron jobs.
- Generally, it's important that, when running upgrade scripts, **nothing else is changing your database**. [There are exceptions, but we don't want to get into them.]

Reply:
> > Its possible that we, or others, might do something out of order / run something manually / need to rerun pieces out of order / etc and having a function that always works vs works based on half of the upgrade script seems safer overall.
> 
> Okay in general, but could you give a specific example of something that we or others might do that would cause my suggested change to malfunction (that wouldn't also cause the existing PR code to malfunction)?

Another site might decide to, for whatever reason, run the contents of the script out of order due to some special something they've done in their DB. Or maybe they had some silent failure in the first half of the script and then the second half runs.

Both methods work. Keeping the additional function seems to me more resilient (and it's already written and works so why change it?).

> Also, if we're recognizing the possibility that someone might need to rerun this script, then it seems like the `ALTER TABLE` and `CREATE TABLE` code at the start should allow for that. Currently, the script will die with an error if you re-execute either of those, won't it?

I'm again thinking of the case where there is some failure somewhere and the admin has to hack the script to pick up where it left off. In that case the user can just comment out the `mysqli_query()` for those and move on.

> Are we allowing for the possibility that a snapshot might be running at the same time as this script? I'm pretty sure we aren't. We should maybe have a comment at the top saying something like: `past_tallies` must remain unchanged over the run (+ possible reruns) of this script, so choose a time well away from any snapshot (or any ad hoc query that could change `past_tallies`).
> 
> On second thought, we should add more stuff to `SETUP/UPGRADE.md` (not necessarily this PR):
> 
>     * Disable general access to your site. (Maybe put it in maintenance mode.)
> 
>     * Disable cron jobs.
> 
>     * Generally, it's important that, when running upgrade scripts, **nothing else is changing your database**. [There are exceptions, but we don't want to get into them.]

Yes, agreed on both the comment in the script and the update to the `UPGRADE.md`.

Reply:
> > [...] could you give a specific example of something that we or others might do that would cause my suggested change to malfunction (that wouldn't also cause the existing PR code to malfunction)?
> 
> Another site might decide to, for whatever reason, run the contents of the script out of order due to some special something they've done in their DB. Or maybe they had some silent failure in the first half of the script and then the second half runs.

To make sure I understand, you're saying if we followed my suggestion (change the script to populate `tally_snapshot_times` before the new columns in `current_tallies` and call `Tallyboard::get_time_of_latest_snapshot` in the latter), and then a site decided to do those parts in the opposite order, and they had already installed the new `Tallyboard.inc`, then `Tallyboard::get_time_of_latest_snapshot` wouldn't work.

Okay, that's true, so I won't push on that suggestion any more.

But it seems like there's no end to things that other sites *might* do. Is there a line beyond which we don't bother to accommodate? Maybe not a well-defined line, but some rough rule of thumb?

> > ... Currently, the script will die with an error if you re-execute either of those, won't it?

> I'm again thinking of the case where there is some failure somewhere and the admin has to hack the script to pick up where it left off.

Ah, what I thought you meant was: the script should be written so that it can be re-run without hacking the script. (It doesn't have to be able to 'pick up where it left off', but it should be able to speed through the stuff that it's already done without error.) I thought that's how some past upgrade scripts had been written.

Reply:
> But it seems like there's no end to things that other sites _might_ do. Is there a line beyond which we don't bother to accommodate? Maybe not a well-defined line, but some rough rule of thumb?

I'm sure there is but I don't know where it is honestly. It would be so much easier to quantify if we knew of a site that was using our code and taking our updates. In the spirit of open source, and because it forces us into some good habits, I want to keep releasing code that others can consume but man does it feel like we're jumping through a lot of hoops for no apparent need.

My minimum bar is "the upgrade scripts should run happy path for users assuming they took the site offline first" but also knowing that we do a lot of special things for pgdp.net that bypass the upgrade scripts`*` may mean that others do too. Indeed its possible that we do some special hacks for pgdp.net in this very same script if the default version takes too long or is too resource intensive. Such as populating the timestamps for S,P1 and then just doing  select insert for (S,P2), (S,P3), etc.

`*` like not converting `page_events` to UTF-8 IIRC.

> Ah, what I thought you meant was: the script should be written so that it can be re-run without hacking the script. (It doesn't have to be able to 'pick up where it left off', but it should be able to speed through the stuff that it's already done without error.) I thought that's how some past upgrade scripts had been written.

We have done that in the past. This one is more challenging because we would have to detect that the additional columns were added to `current_tallies` and skip that update. The `CREATE TABLE` could be done with the `IF NOT EXISTS` but there's no workaround for the alter that I'm aware of.

Reply:
> My minimum bar is "the upgrade scripts should run happy path for users assuming they took the site offline first" ...

Is it fair to state that outright in the interest of simplifying things? (Or would it?) I know it's going to be a complex upgrade, so, hopefully anyone using our code should, also.

Reply:
Every variation that Michael and I have talked about will work for the happy path. At this point we're mostly bikeshedding 😁